File: predecessors/scraper-backend-python/src/payments/subscription/firestore.py
class FirestoreManagerSubscription():
    def __init__(self, db):
        self.db = db
        
    def create(self, subscription_data):
        # Create a new subscription document in Firestore
        doc_ref = self.db.collection('subscriptions').document()
        doc_ref.set(subscription_data)
        return doc_ref.id

    def get(self, subscription_id):
        # Retrieve subscription data from Firestore
        doc_ref = self.db.collection('subscriptions').document(subscription_id)
        doc = doc_ref.get()

        if doc.exists:
            return doc.to_dict()

        return None

    def update(self, subscription_id, subscription_data):
        # Update an existing subscription document in Firestore
        doc_ref = self.db.collection('subscriptions').document(subscription_id)
        doc_ref.update(subscription_data)

    def delete(self, subscription_id):
        # Delete a subscription document from Firestore
        doc_ref = self.db.collection('subscriptions').document(subscription_id)
        doc_ref.delete()

File: predecessors/scraper-backend-go/internal/payments/subscriptions/firestore.go
package subscriptions

import (
	"context"

	"cloud.google.com/go/firestore"
)

type FirestoreSubscriptionManager struct {
	db *firestore.Client
}

func NewFirestoreSubscriptionManager(client *firestore.Client) *FirestoreSubscriptionManager {
	return &FirestoreSubscriptionManager{
		db: client,
	}
}

func (manager *FirestoreSubscriptionManager) Create(subscription Subscription) (string, error) {
	docRef, _, err := manager.db.Collection("subscriptions").Add(context.Background(), subscription)
	if err != nil {
		return "", err
	}

	return docRef.ID, nil
}

func (manager *FirestoreSubscriptionManager) Get(subscriptionID string) (*Subscription, error) {
	docRef := manager.db.Collection("subscriptions").Doc(subscriptionID)
	doc, err := docRef.Get(context.Background())
	if err != nil {
		if doc == nil {
			return nil, nil
		}
		return nil, err
	}

	var subscription Subscription
	err = doc.DataTo(&subscription)
	if err != nil {
		return nil, err
	}

	subscription.ID = docRef.ID
	return &subscription, nil
}

func (manager *FirestoreSubscriptionManager) Update(subscriptionID string, subscription Subscription) error {
	docRef := manager.db.Collection("subscriptions").Doc(subscriptionID)
	_, err := docRef.Set(context.Background(), subscription, firestore.MergeAll)
	return err
}

func (manager *FirestoreSubscriptionManager) Delete(subscriptionID string) error {
	docRef := manager.db.Collection("subscriptions").Doc(subscriptionID)
	_, err := docRef.Delete(context.Background())
	return err
}
